source/common/stream_info/filter_state_impl.cc

```diff
@@ -5,8 +5,26 @@
 namespace Envoy {
 namespace StreamInfo {
 
-void FilterStateImpl::setData(absl::string_view data_name, std::unique_ptr<Object>&& data,
-                              FilterState::StateType state_type) {
+FilterStateImpl::FilterStateImpl(std::shared_ptr<FilterState> parent,
+                                 FilterState::LifeSpan life_span)
+    : parent_(parent), life_span_(life_span) {
+  if (life_span >= FilterState::LifeSpan::TopSpan) {
+    parent_ = nullptr;
+    return;
+  }
+  // Recursively create parent FilterState if no parent is provided or parent is not the immediate
+  // parent.
+  if (parent_ == nullptr || parent_->life_span() != life_span_ + 1) {
+    parent_ = std::make_shared<FilterStateImpl>(parent_, FilterState::LifeSpan(life_span_ + 1));
+  }
+}
+
+void FilterStateImpl::setData(absl::string_view data_name, std::shared_ptr<Object> data,
```

Comment:
Q: Do we want to somehow prevent "shadowing" data? Should we at least ASSERT on this? I could see this being a very confusing programming error.

Reply:
Added code to throw exception.